runtime/doc/msgpack_rpc.txt

```diff
@@ -241,4 +241,152 @@ Even for statically compiled clients it is good practice to avoid hardcoding
 the type codes, because a client may be built against one Nvim version but
 connect to another with different type codes.
 
+==============================================================================
+6. Buffer Updates				                *buffer-updates* *rpc-buffer-updates*
+
+A dedicated API has been created to allow co-processes to be notified  when a
+buffer is changed in any way. It is difficult and error-prone to try and do
+this with autocommands such as |TextChanged|.
+
+                                                           *buffer-updates-events*
+BufferUpdates Events~
+
+The co-process will start receiving the following notification events:
+
+nvim_buf_updates_start[{buf}, {changedtick}, {linedata}, {more}]	 *nvim_buf_updates_start*
+
+  Nvim will send at least one of these notifications to confirm that buffer
+  updates are registered for this plugin, and possibly send the buffer's
+  contents. If the buffer is very large, nvim might send the contents
+  through in multiple events to avoid loading the entire buffer's contents
+  into memory at once.
+
+	{buf} is an API handle for the buffer.
+
+	{changedtick} is the value of |b:changedtick| for the buffer. If you
+	send an API command back to nvim you can check the value of
+	|b:changedtick| as part of your request to ensure that no other
+	changes have been made. 
+
+  {linedata} is a list of strings containing the buffer's contents. If this
+  list contains 100 strings, then they represent lines 1-100 of the buffer.
+  Newline characters are not included in the strings, so empty lines will be
+  given as empty strings. If you receive another |nvim_buf_updates_start|
+  notification with another {linedata} list, then these lines represent the
+  next N lines of the buffer. I.e., a second notification with another list of
+  100 strings will represent lines 101-200 of the buffer. If you send the
+  |nvim_buf_updates_start| request with its argument set to `"False"`, this
+  will be empty.
+
+  {linedata} will always have at least 1 item, but the maximum length is
+  determined by nvim and not guaranteed to be any particular size.  Also the
+  number of {linedata} items may vary between notifications, so your plugin
+  must be prepared to receive the line data in whatever size lists nvim
+  decides to split it into.
+
+  {more} is a boolean which tells you whether or not to expect more
+  |nvim_buf_updates_start| notifications. When {more} is false, you can be certain
+  that you now have the entire buffer's contents.
```

Comment:
Why don't we use this same mechanism for `nvim_buf_update`? And why is a separate `nvim_buf_updates_start` event needed?

Reply:
Not needed. I agree it would be nicer to just send `nvim_buf_update` for everything, and have the `{more}` parameter also for non-initial updates, even if we don't use it yet. #workingonit

Reply:
As a user of this feature, I'm not particularly thrille having to specialcase for `lastline=-1`. Otoh, I specialised for the event before, so it's not a huge deal.

Just to be clear, if a non-initial sending includes the last line or even the whole buffer, we still send "real" numbers, not -1, right?

Reply:
Well, previously you had to specialcase an entire separate event.

> Just to be clear, if a non-initial sending includes the last line or even the whole buffer, we still send "real" numbers, not -1, right?

Yes. In that case you are still replacing a range of lines that the client already knows about, which could happen to be the previous entire buffer, so this should still use the concrete length. But with the first send there is no previous content for the client to replace (it knows nothing, or rather nvim doesn't know if it knows anything ), so specifying a concrete length doesn't make sense. Rather `[0, -1]` would mean "regardless of previous state of information (probably nothing), these lines are the entire buffer".

Reply:
@KillTheMule fair point, but clients already would have to handle `nvim_buf_updates_start` as a special case.

The semantics of `-1` are well established by e.g `nvim_buf_{get,set}_lines`.

Reply:
Actually, do we need a `{more}` parameter? You always need to expect ~spanish~ buffer updates. So if you request the buffer's content, and get something that's not the whole buffer, the next update will get you up-to-date, the line numbers will tell you specifics.

Right?

That of course means that we can't use `-1` for the initial send (at least if it's not the whole buffer), but I don't think that's really needed for anything.

Reply:
> That of course means that we can't use -1 for the initial send (at least if it's not the whole buffer), but I don't think that's really needed for anything.

I don't think that follows. The first send still starts from an unknown state, but the following (even if they help build the "true" initial state) builds from the incomplete but concrete client state.

But I still think we want to somehow separate "real" buffer states, from ones invented for transport technical reasons. How about setting changedtick to `-1` in the cases that would have `more`, if we want to avoid this flag? Because these do not correspond to any real buffer state, while true buffer states are identified by their changedtick.

Reply:
>  One can identify those because the changedtick stays the same, if need be, but I don't really see why that would be neccessary.

No you cannot, as you when you receive one event don't yet know if more events will come with the same changetick. But if separate `more` flag (which I imagined would be usable _beyond_ the initial send, for similarly large updates) is preferred, let's use that :)

Reply:
> which I imagined would be usable beyond the initial send, for similarly large updates

That's what I was thinking, seems like `more` ties in with the "chunking" discussion near the top of this PR.